dsd/constraints.py

```diff
@@ -616,17 +652,45 @@ def generate_sequence(self, rng: np.random.Generator) -> str:
 
         :param rng:
             numpy random number generator to use. To use a default, pass :py:data:`np.default_rng`.
+        :param previous_sequence:
+            previously generated sequence to be replaced by a new sequence; optional if no previous
+            sequence exists.
         :return:
             DNA sequence of given length satisfying :py:data:`DomainPool.numpy_constraints` and
             :py:data:`DomainPool.sequence_constraints`
         """
         log_debug_sequence_constraints_accepted = False
-        sequence = self._get_next_sequence_satisfying_numpy_constraints(rng)
-        while not self.satisfies_sequence_constraints(sequence):
-            logger.debug(f'rejecting domain sequence {sequence}; failed some sequence constraint')
+        if not self.find_neighbor_sequences or not previous_sequence:
+            # takes a completely random sequence from domain pool
             sequence = self._get_next_sequence_satisfying_numpy_constraints(rng)
-        if log_debug_sequence_constraints_accepted:
-            logger.debug(f'accepting domain sequence {sequence}; passed all sequence constraints')
+            while not self.satisfies_sequence_constraints(sequence):
+                logger.debug(f'rejecting domain sequence {sequence}; failed some sequence constraint')
+                sequence = self._get_next_sequence_satisfying_numpy_constraints(rng)
+            if log_debug_sequence_constraints_accepted:
+                logger.debug(f'accepting domain sequence {sequence}; passed all sequence constraints')
+        else:
```

Comment:
This seems inefficient. It looks like the Hamming graph is freshly constructed each time `generate_sequence` is called.

The way I pictured this working, the graph would be constructed once when the sequences are generated, at the end of the method `_generate_sequences_satisfying_numpy_constraints`. Then the graph is available to use already, each time `generate sequence` is called, until we run out of sequences. Each time `generate_sequence` is called, at the end, it should remove `previous sequence` from the graph.